commons-core/src/main/scala/com/avsystem/commons/serialization/cbor/CborAdtMetadata.scala

```diff
@@ -0,0 +1,245 @@
+
+      case _ =>
+        codec
+    }
+
+    def validate(): Unit = {
+      cases.groupBy(_.keyInfo.rawKey).foreach {
+        case (_, List(_)) =>
+        case (_, multipleCases) =>
+          val caseNames = multipleCases.map(_.keyInfo.sourceName).mkString(", ")
+          throw new Exception(s"case classes/objects $caseNames in sealed hierarchy $sourceName have the same CBOR key")
+      }
+      cases.foreach(_.ensureUniqueKeys(discriminator.map(_.rawKey)))
+    }
+  }
+
+  sealed trait Case[T] extends CborAdtMetadata[T] {
+    def keyInfo: CborKeyInfo[T]
+    def ensureUniqueKeys(discriminatorKey: Opt[RawCbor]): Unit
+  }
+
+  @positioned(positioned.here)
+  final class Record[T](
+    @composite val keyInfo: CborKeyInfo[T],
+    @multi @adtParamMetadata val fields: List[Field[_]]
+  ) extends Case[T] {
+    private val keyCodec = new MappingCborKeyCodec(fields.map(_.keyInfo))
+
+    def adjustCodec(codec: GenObjectCodec[T]): GenObjectCodec[T] =
+      new CborRawKeysCodec(codec, keyCodec)
+
+    def adjustFlatCaseCodec(codec: OOOFieldsObjectCodec[T]): OOOFieldsObjectCodec[T] =
+      new OOOFieldCborRawKeysCodec[T](codec, keyCodec)
+
+    def ensureUniqueKeys(discriminatorKey: Opt[RawCbor]): Unit = {
+      for (disc <- discriminatorKey; field <- fields) {
+        if (disc == field.keyInfo.rawKey) {
+          throw new Exception(s"field ${field.keyInfo.sourceName} in case class ${keyInfo.sourceName}" +
+            s" has the same CBOR key as the discriminator key")
+        }
+      }
+      fields.groupBy(_.keyInfo.rawKey).foreach {
+        case (_, List(_)) =>
+        case (_, multipleFields) =>
+          val fieldNames = multipleFields.map(_.keyInfo.sourceName).mkString(", ")
+          throw new Exception(s"fields $fieldNames in case class ${keyInfo.sourceName} have the same CBOR key")
+      }
+    }
+
+    def validate(): Unit = ensureUniqueKeys(Opt.Empty)
+  }
+
+  final class Field[T](
+    @composite val keyInfo: CborKeyInfo[T]
+  ) extends TypedMetadata[T]
+
+  @positioned(positioned.here)
+  final class Singleton[T](
+    @infer @checked val value: ValueOf[T],
+    @composite val keyInfo: CborKeyInfo[T],
+  ) extends Case[T] {
+    def adjustCodec(codec: GenObjectCodec[T]): GenObjectCodec[T] = codec
+    def ensureUniqueKeys(discriminatorKey: Opt[RawCbor]): Unit = ()
+    def validate(): Unit = ()
+  }
+
+  private class MappingCborKeyCodec(keyInfos: List[CborKeyInfo[_]]) extends CborKeyCodec {
+    val rawKeys: Map[String, RawCbor] =
+      keyInfos.mkMap(_.stringKey, _.rawKey)
+    val strKeys: Map[RawCbor, String] =
+      keyInfos.mkMap(_.rawKey, _.stringKey)
+
+    def writeFieldKey(fieldName: String, output: CborOutput): Unit =
+      rawKeys.getOpt(fieldName).fold(output.writeString(fieldName))(output.writeRawCbor)
+    def readFieldKey(input: CborInput): String =
+      input.readRawCbor() |> (cbor => strKeys.getOrElse(cbor, s"unknown_cbor_key:$cbor")) // unknown key will be ignored
```

Comment:
unknown String keys are ignored as well, right?
Is there a flag to fail on unrecognized string fields? If yes, does it apply to CBOR as well?

Reply:
There is no such feature in `GenCodec` itself. Unknown fields are ignored so that case classes may evolve in a backwards compatible way.